plugins/cryptolib/cryptolib.cpp

```diff
@@ -0,0 +1,669 @@
+// PUBLIC KEY SUPPORT
+//-------------------------------------------------------------------------------------------------------------------------------------------
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clInstalledPublicKeyAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    loadDigests();
+#endif
+    OBJ_NAME_do_all_sorted(OBJ_NAME_TYPE_PKEY_METH, my_clAlgorithmsCallback, (void*)&algorithms);
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clSupportedPublicKeyAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+    algorithms.appendList("SHA256", ",");
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+//signature helper function
+void doPKISign(size32_t & __lenResult, void * & __result,
+        IDigitalSignatureManager * pDSM,
+        size32_t lenInputdata, const void * inputdata)
+{
+    StringBuffer sbSig;
+    if (pDSM->digiSign(sbSig, lenInputdata, inputdata))
+    {
+        __result = CTXMALLOC(parentCtx, sbSig.length());
+        __lenResult = sbSig.length();
+        memcpy(__result, sbSig.str(), __lenResult);
+    }
+    else
+    {
+        throw MakeStringException(-1, "Unable to create Digital Signature");
+    }
+}
+
+void verifyPKIAlgorithm(const char * pkAlgorithm)
+{
+    if (!strieq(pkAlgorithm, "SHA256"))
+        throw MakeStringException(-1, "Unsupported PKI algorithm (%s) specified", pkAlgorithm);
+}
+
+
+//-----------------------------------------------------------------
+//  Simple cache for instances of Digital Signature Managers
+//-----------------------------------------------------------------
+class CDSMCache
+{
+private:
+    CriticalSection signingDSMCacheCS;//guards modifications to the cache map
+    CriticalSection verifyingDSMCacheCS;//guards modifications to the cache map
+    typedef std::unordered_map <string, IDigitalSignatureManager *> DSMCache;//DSM instances are never deleted, since we are not notified if a module is destroyed
+    DSMCache signingDSMCache;
+    DSMCache verifyingDSMCache;
+
+public:
+    ~CDSMCache()
+    {
+        for ( auto it = signingDSMCache.begin(); it != signingDSMCache.end(); ++it )
+        {
+            ((IInterface *)(*it).second)->Release();
+        }
+        for ( auto it = verifyingDSMCache.begin(); it != verifyingDSMCache.end(); ++it )
+        {
+            ((IInterface *)(*it).second)->Release();
+        }
+    }
+
+    IDigitalSignatureManager * querySigningInstance(const char * algo, const char * privKeyFS, const char * privKeyBuff, const char * passphrase)
+    {
+        VStringBuffer searchKey("%s_%s_%s_%s", algo, privKeyFS, privKeyBuff, passphrase);
```

Comment:
FS or Buff could be null, and %s behaviour is undefined (could crash).
So should build up checking conditionally.